File: client/cloudflareturn.go
// Cloudflare Calls TURN integration.
//
// Cloudflare offers a globally-distributed TURN service ($0.05/GB
// outbound, free for the first 1 TB/month). Unlike most TURN
// providers, credentials are short-lived: the user's browser/app
// authenticates with Cloudflare's API using a long-lived TOKEN_ID +
// API_TOKEN pair, and Cloudflare returns ICE-server credentials
// valid for a configurable window (default 1 hour).
//
// We persist the long-lived pair in SQLite. On every CreatePortal /
// JoinPortal we fetch fresh creds and inject them into the WebRTC
// configuration before constructing the mesh.Manager.
//
// Docs: https://developers.cloudflare.com/calls/turn/
package main

import (
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"sync"
	"time"

	"github.com/pion/webrtc/v4"
	"portal_traffic_client/mesh"
)

// mesh_DefaultICEServers exposes mesh's default servers for our
// resolveICEServers helper without circular imports.
var mesh_DefaultICEServers = mesh.DefaultICEServers

// CloudflareTurnConfig is the long-lived API credentials.
type CloudflareTurnConfig struct {
	TokenID  string `json:"tokenId"`
	APIToken string `json:"apiToken"`
}

// cloudflareTurnCache holds the most recently fetched short-lived
// credentials so we don't burn a Cloudflare API call on every
// CreatePortal/JoinPortal. Cached for 30 minutes — well under the
// 1-hour default TTL.
type cloudflareTurnCache struct {
	mu      sync.Mutex
	servers []webrtc.ICEServer
	until   time.Time
}

// cfTurnGenResponse mirrors the Cloudflare API response shape.
type cfTurnGenResponse struct {
	IceServers struct {
		URLs       interface{} `json:"urls"`
		Username   string      `json:"username"`
		Credential string      `json:"credential"`
	} `json:"iceServers"`
}

// fetchCloudflareTurn calls Cloudflare's credentials/generate
// endpoint with the configured token pair and returns ICE servers
// the WebRTC engine can use. ttlSeconds=3600 (1h) is the default.
//
// Errors are returned for: missing config, API failure, malformed
// response. The caller should fall back to default ICE servers.
func fetchCloudflareTurn(ctx context.Context, cfg CloudflareTurnConfig, ttlSeconds int) ([]webrtc.ICEServer, error) {
	if cfg.TokenID == "" || cfg.APIToken == "" {
		return nil, errors.New("cloudflare turn: token_id va api_token to'liq emas")
	}
	if ttlSeconds <= 0 {
		ttlSeconds = 3600
	}

	body, _ := json.Marshal(map[string]int{"ttl": ttlSeconds})
	url := fmt.Sprintf(
		"https://rtc.live.cloudflare.com/v1/turn/keys/%s/credentials/generate",
		cfg.TokenID,
	)
	req, err := http.NewRequestWithContext(ctx, "POST", url, bytes.NewReader(body))
	if err != nil {
		return nil, err
	}
	req.Header.Set("Authorization", "Bearer "+cfg.APIToken)
	req.Header.Set("Content-Type", "application/json")

	client := &http.Client{Timeout: 8 * time.Second}
	resp, err := client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("cloudflare turn request: %w", err)
	}
	defer resp.Body.Close()
	raw, _ := io.ReadAll(resp.Body)
	if resp.StatusCode != 200 {
		return nil, fmt.Errorf("cloudflare turn HTTP %d: %s", resp.StatusCode, snippet(raw))
	}
	var parsed cfTurnGenResponse
	if err := json.Unmarshal(raw, &parsed); err != nil {
		return nil, fmt.Errorf("cloudflare turn parse: %w", err)
	}

	urls := normaliseURLs(parsed.IceServers.URLs)
	if len(urls) == 0 {
		return nil, errors.New("cloudflare turn: no urls in response")
	}
	return []webrtc.ICEServer{
		{
			URLs:       urls,
			Username:   parsed.IceServers.Username,
			Credential: parsed.IceServers.Credential,
		},
	}, nil
}

// normaliseURLs accepts the Cloudflare response shape: either a
// single string or a string slice. Returns []string.
func normaliseURLs(v interface{}) []string {
	switch t := v.(type) {
	case string:
		return []string{t}
	case []interface{}:
		out := make([]string, 0, len(t))
		for _, x := range t {
			if s, ok := x.(string); ok {
				out = append(out, s)
			}
		}
		return out
	}
	return nil
}

func snippet(b []byte) string {
	if len(b) > 200 {
		b = b[:200]
	}
	return string(b)
}

// ----------------------------------------------------------------------------
// App methods
// ----------------------------------------------------------------------------

// GetCloudflareTurn returns the persisted Cloudflare TURN config
// (token id + API token). Empty values mean unconfigured.
func (a *App) GetCloudflareTurn() CloudflareTurnConfig {
	a.mu.RLock()
	store := a.store
	a.mu.RUnlock()
	if store == nil {
		return CloudflareTurnConfig{}
	}
	return CloudflareTurnConfig{
		TokenID:  store.GetOr("cf_turn_token_id", ""),
		APIToken: store.GetOr("cf_turn_api_token", ""),
	}
}

// SetCloudflareTurn persists the Cloudflare TURN credentials.
// Either field empty disables Cloudflare TURN; both must be set
// for it to be used.
func (a *App) SetCloudflareTurn(cfg CloudflareTurnConfig) error {
	a.mu.RLock()
	store := a.store
	a.mu.RUnlock()
	if store == nil {
		return errors.New("storage mavjud emas")
	}
	_ = store.PutSetting("cf_turn_token_id", cfg.TokenID)
	_ = store.PutSetting("cf_turn_api_token", cfg.APIToken)
	// Invalidate any cached credentials so the next call refetches.
	a.cfTurnCache.mu.Lock()
	a.cfTurnCache.until = time.Time{}
	a.cfTurnCache.servers = nil
	a.cfTurnCache.mu.Unlock()
	return nil
}

// TestCloudflareTurn calls the API and reports whether the creds work,
// what URLs / username came back, and how long the call took. Lets
// the user verify their setup without creating a portal.
func (a *App) TestCloudflareTurn() TurnTestResult {
	cfg := a.GetCloudflareTurn()
	res := TurnTestResult{}
	if cfg.TokenID == "" || cfg.APIToken == "" {
		res.Message = "Cloudflare TURN sozlanmagan (Settings → Cloudflare TURN)"
		return res
	}

	start := time.Now()
	ctx, cancel := context.WithTimeout(context.Background(), 8*time.Second)
	defer cancel()
	servers, err := fetchCloudflareTurn(ctx, cfg, 600)
	res.GatherMs = time.Since(start).Milliseconds()
	if err != nil {
		res.Message = err.Error()
		a.logger.Warn("cloudflare turn test failed", "err", err)
		return res
	}
	for _, s := range servers {
		res.URLs = append(res.URLs, s.URLs...)
	}
	res.OK = true
	res.HadRelay = true
	res.Types = []string{"relay"}
	res.Message = "Cloudflare TURN ishlamoqda — short-lived credentials qaytarildi ✓"
	a.logger.Info("cloudflare turn test ok", "urls", res.URLs)
	return res
}

// resolveICEServers builds the final ICE-server list for a session.
// Precedence:
//  1. Cloudflare TURN (if configured + credentials fetch succeeds)
//  2. Manual TURN URL (if configured)
//  3. Default STUN-only fallback
//
// Cached Cloudflare creds are reused for 30 minutes to minimise API
// calls.
func (a *App) resolveICEServers() []webrtc.ICEServer {
	cf := a.GetCloudflareTurn()
	if cf.TokenID != "" && cf.APIToken != "" {
		// Try cache first.
		a.cfTurnCache.mu.Lock()
		valid := time.Now().Before(a.cfTurnCache.until) && len(a.cfTurnCache.servers) > 0
		cached := a.cfTurnCache.servers
		a.cfTurnCache.mu.Unlock()
		if valid {
			a.logger.Debug("cloudflare turn cache hit")
			return append(append([]webrtc.ICEServer{}, mesh_DefaultICEServers...), cached...)
		}
		ctx, cancel := context.WithTimeout(context.Background(), 8*time.Second)
		defer cancel()
		servers, err := fetchCloudflareTurn(ctx, cf, 3600)
		if err == nil && len(servers) > 0 {
			a.cfTurnCache.mu.Lock()
			a.cfTurnCache.servers = servers
			a.cfTurnCache.until = time.Now().Add(30 * time.Minute)
			a.cfTurnCache.mu.Unlock()
			a.logger.Info("cloudflare turn fetched", "url_count",
				countServerURLs(servers))
			return append(append([]webrtc.ICEServer{}, mesh_DefaultICEServers...), servers...)
		}
		a.logger.Warn("cloudflare turn fetch failed; falling back", "err", err)
	}
	// No Cloudflare TURN — let the mesh layer apply manual TurnURL
	// from settings (handled inside mesh.New) on top of the STUN
	// defaults. Returning nil signals "use mesh defaults + manual".
	return nil
}

func countServerURLs(servers []webrtc.ICEServer) int {
	n := 0
	for _, s := range servers {
		n += len(s.URLs)
	}
	return n
}

File: mobile/app/src/main/java/uz/aihealth/portal_mobile/turn/CloudflareTurn.kt
// Cloudflare Calls TURN client. Mirrors client/cloudflareturn.go on the
// desktop: one POST to credentials/generate using a long-lived TOKEN_ID
// + API_TOKEN, returning short-lived ICE-server creds.
//
// We cache the result for 30 minutes (well under the 1h default TTL)
// because hitting the API on every portal create / join is wasteful
// and adds 100-300ms of latency to the welcome→ready transition.
//
// Docs: https://developers.cloudflare.com/calls/turn/
//
// Threading: all entry points are suspend; HTTP runs on Dispatchers.IO
// inside [withContext]. The cache is a synchronized object — call
// volume is too low for atomic primitives to matter.

package uz.aihealth.portal_mobile.turn

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.coroutines.withTimeoutOrNull
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonObject
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.webrtc.PeerConnection
import java.util.concurrent.TimeUnit

private const val TAG = "CloudflareTurn"
private const val ENDPOINT = "https://rtc.live.cloudflare.com/v1/turn/keys/%s/credentials/generate"

object CloudflareTurn {

    private val http = OkHttpClient.Builder()
        .connectTimeout(8, TimeUnit.SECONDS)
        .readTimeout(8, TimeUnit.SECONDS)
        .build()

    private val lenientJson = Json { ignoreUnknownKeys = true; isLenient = true }

    /**
     * Single cache slot. Cloudflare creds are scoped to (tokenId), so
     * we store the tokenId we minted them with — if the user changes
     * their config we'll naturally invalidate without an explicit
     * `.invalidate()` call.
     */
    private data class CacheEntry(
        val tokenId: String,
        val server: PeerConnection.IceServer,
        val expiresAtMs: Long,
    )

    @Volatile private var cache: CacheEntry? = null

    /**
     * Fetch fresh ICE-server creds. Returns null on any failure — the
     * caller should fall back to STUN-only / manual TURN.
     *
     * [ttlSeconds] becomes the credential lifetime Cloudflare hands us;
     * 3600 (1h) matches the desktop default and the API's own default.
     */
    suspend fun fetch(cfg: CloudflareTurnConfig, ttlSeconds: Int = 3600): PeerConnection.IceServer? {
        if (!cfg.isConfigured()) return null

        cache?.let { c ->
            if (c.tokenId == cfg.tokenId && System.currentTimeMillis() < c.expiresAtMs) {
                return c.server
            }
        }

        val server = doFetch(cfg, ttlSeconds) ?: return null
        // Cache for 30 minutes regardless of the TTL Cloudflare claims;
        // if we cached for the full hour and the user toggled the config
        // we'd still serve stale creds for 60min.
        cache = CacheEntry(
            tokenId = cfg.tokenId,
            server = server,
            expiresAtMs = System.currentTimeMillis() + 30 * 60 * 1_000L,
        )
        return server
    }

    /**
     * Settings → "Sinash" button. Bypasses cache and reports whether
     * the API actually answered, what URLs came back, and how long it
     * took — enough info for the user to debug a typo without leaving
     * the screen.
     */
    suspend fun test(cfg: CloudflareTurnConfig): TurnTestResult {
        if (!cfg.isConfigured()) {
            return TurnTestResult(ok = false, message = "Cloudflare TURN: Token ID + API Token to'liq emas")
        }
        val start = System.currentTimeMillis()
        val server = withTimeoutOrNull(8_000L) { doFetch(cfg, ttlSeconds = 600) }
        val elapsed = System.currentTimeMillis() - start
        return if (server == null) {
            TurnTestResult(ok = false, message = "Cloudflare TURN xato — credentials olinmadi", gatherMs = elapsed)
        } else {
            TurnTestResult(
                ok = true,
                message = "Cloudflare TURN ishlamoqda — short-lived credentials qaytarildi ✓",
                urls = server.urls.toList(),
                gatherMs = elapsed,
            )
        }
    }

    fun invalidateCache() {
        cache = null
    }

    private suspend fun doFetch(cfg: CloudflareTurnConfig, ttlSeconds: Int): PeerConnection.IceServer? =
        withContext(Dispatchers.IO) {
            val body = """{"ttl": $ttlSeconds}""".toRequestBody("application/json".toMediaType())
            val req = Request.Builder()
                .url(ENDPOINT.format(cfg.tokenId))
                .post(body)
                .header("Authorization", "Bearer ${cfg.apiToken}")
                .header("Content-Type", "application/json")
                .build()

            runCatching {
                http.newCall(req).execute().use { resp ->
                    val raw = resp.body?.string().orEmpty()
                    if (!resp.isSuccessful) {
                        android.util.Log.w(TAG, "HTTP ${resp.code}: ${raw.take(200)}")
                        return@use null
                    }
                    parse(raw)
                }
            }.getOrElse { e ->
                android.util.Log.w(TAG, "fetch failed: ${e.message}")
                null
            }
        }

    private fun parse(raw: String): PeerConnection.IceServer? {
        val obj = runCatching { lenientJson.parseToJsonElement(raw).jsonObject }.getOrNull() ?: return null
        val ice = obj["iceServers"] as? JsonObject ?: return null
        val urls = normaliseUrls(ice["urls"]) ?: return null
        if (urls.isEmpty()) return null
        val username = (ice["username"] as? JsonPrimitive)?.contentOrNull.orEmpty()
        val credential = (ice["credential"] as? JsonPrimitive)?.contentOrNull.orEmpty()
        return PeerConnection.IceServer.builder(urls)
            .setUsername(username)
            .setPassword(credential)
            .createIceServer()
    }

    /** Cloudflare may return urls as a single string or an array of strings. */
    private fun normaliseUrls(v: JsonElement?): List<String>? = when (v) {
        null -> null
        is JsonPrimitive -> v.contentOrNull?.let { listOf(it) }
        is JsonArray -> v.mapNotNull { (it as? JsonPrimitive)?.contentOrNull }
        else -> null
    }
}
